rfd/0017-supply-chain-hardening.md

```diff
@@ -0,0 +1,59 @@
+# Supply Chain Hardening
+
+To prevent supply chain attacks take control over all software distribution channels for Teleport and 
+only distribute signed software.
+
+## Why 
+
+Have a cryptographicaly-verifiable audit trail of all software changes and releases 
+and protect all software distribution channels.
+
+## How
+
+### Github
+
+Require [signed commits](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification) 
+for protected branches `master` and `branch/x.y`.
+
+Turn on commit signing for users and bots following steps 
+[here](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/about-commit-signature-verification)
+
+**Github Mirror**
+
+Self host git repository for Teleport OSS, Enterprise, Cloud and webapps repositories.
+To allow outside collaboration, allow pull requests, forks, but process all commits using CI/CD pipeline 
+and do not use Github as a source of truth.
+
+**Teleport specific key repository**
+
+Establish Teleport-specific repository of supported public keys used by core team developers and have an 
+additional verification step done by Drone pipeline.
+
+Developers should register [Yubikeys](https://www.engineerbetter.com/blog/yubikey-signed-commits/)
+as used keys.
+
+**Support for external contributors**
+
+Once pull request from external contributor has been approved, C/CD can amend the commit, sign it using our CI/CD key
+and add to master.
+
+**Remove the github releases files**
+
+Remove the source releases files generated by Github and instead publish Drone-generated source files alongside the signatures and hashes.
```

Comment:
Sorry, it's not clear to me what this does or the security properties that change?

Reply:
The context is [this issue](https://github.com/gravitational/teleport/issues/5174) where it appears that the behaviour of Github's automatic source tarball generation code has changed, resulting in the SHA256 hashes of the source archives changing without commits from our end. Controlling the process of generating the source tarballs ourselves removes a dependency on Github's code (which we have no control of)

Reply:
>  resulting in the SHA256 hashes of the source archives changing without commits from our end

Are we actually sure about this, or just assuming it's the case because we don't know why it changed on that particular asset. And don't have a copy of the old asset to compare against, but presumably compared to the sources and the asset appears legitimate? And we're sure the asset actually did change, and not that the reporter accidentally hashed a partial download or something.

I think my advice would be if we're looking at signing the commits anyways, is to shallow copy of the repo and validate the commits with GPG. Because otherwise we're introducing new key handling to sign source files, and methods to communicate those keys to customers who want to validate signatures, and then need to be concerned with rotating those secrets, storing them, etc, etc. Not that GPG signed commits don't have all those same problems, it's just double the work to do it for the sources in git, and to do it again with CI/CD and the assets tarball. 

Although I suppose it would work to just have drone shallow copy the tag and overwrite the github generates releases, so all the signature are still in place. I don't know if you can fully validate the git history with just a shallow copy though. No new key handling or security.

Reply:
I downloaded the archive, unpacked it and the contents are a bit different: 

```bash

$ diff -r teleport-5.0.2 teleport

diff -r teleport-5.0.2/vendor/k8s.io/client-go/pkg/version/base.go teleport/vendor/k8s.io/client-go/pkg/version/base.go
58,59c58,59
< gitVersion   string = "v0.0.0-master+17679dcec2"
< gitCommit    string = "17679dcec24796955b96485d5922b6d695735a3b" // sha1 from git, output of $(git rev-parse HEAD)
---
> gitVersion   string = "v0.0.0-master+$Format:%h$"
> gitCommit    string = "$Format:%H$" // sha1 from git, output of $(git rev-parse HEAD)
```

It may as well be user/CI/CD or some other error. I'm not as concerned about this particular use-case though. I'm concerned about the fact that we have no way of verifying the integrity of the sources against the git commits.

Reply:
The fact that it's happened for many of our source archives at once (we checked 4.4.4, 4.4.5, 5.0.0, 5.0.1 and 5.0.2 - all are affected) points to an upstream issue.

There are just two differences in the tarball versus the tag checkout - `%h` and `%H` in a vendor file were replaced with actual commit hashes. It looks a lot like a bug.